src/main/java/org/interledger/InterledgerAddress.java

```diff
@@ -109,12 +133,151 @@ default boolean startsWith(final InterledgerAddress interledgerAddress) {
+
+      private static final String REGEX = "(?=^.{1,1023}$)"
+          + "^(g|private|example|peer|self|test[1-3])[.]([a-zA-Z0-9_~-]+[.])*([a-zA-Z0-9_~-]+)?$";
+
+      private static final Pattern PATTERN = Pattern.compile(REGEX);
+
+      private final String value;
+
+      /**
+       * Required-args Constructor.
+       *
+       * @param builder An non-null instance of {@link Builder} to construct a new instance of.
+       */
+      private Impl(final Builder builder) {
+        Objects.requireNonNull(builder, "Builder must not be null!");
+        Objects.requireNonNull(builder.value, "value must not be null!");
+
+        if (!isValidInterledgerAddress(builder.value)) {
+          throw new IllegalArgumentException(
+              "Invalid characters in address.  Reference Interledger RFC-15 for proper format.");
+        }
+
+        this.value = builder.value;
+      }
+
+      /**
+       * Helper method to determine if an Interledger Address conforms to the specifications
+       * outlined in Interledger RFC #15.
+       *
+       * @param value A {@link String} representing a potential Interledger Address value.
+       *
+       * @return {@code true} if the supplied {@code value} conforms to the requirements of RFC 15;
+       * {@code false} otherwise.
+       *
+       * @see "https://github.com/interledger/rfcs/tree/master/0015-ilp-addresses"
+       */
+      private boolean isValidInterledgerAddress(final String value) {
+        Objects.requireNonNull(value);
+        return PATTERN.matcher(value)
+            .matches();
+      }
+
+      /**
+       * Accessor method for this address's {@link String} value. <p> NOTE: This is distinct of
+       * {@link #toString()} to allow for the two values to diverge, e.g., for debugging or logging
+       * purposes. </p>
+       *
+       * @return The value of the {@code value} attribute
+       */
+      @Override
+      public String getValue() {
+        return this.value;
+      }
+
+      @Override
+      public InterledgerAddress with(String segment) {
+        Objects.requireNonNull(segment, "Segment String must not be null!");
+        final StringBuilder sb = new StringBuilder(this.getValue());
+        if (!this.isLedgerPrefix()) {
+          sb.append(".");
+        }
+        sb.append(segment);
+
+        return new Builder().value(sb.toString())
+            .build();
+      }
+
+      @Override
+      public boolean equals(final Object object) {
+        if (this == object) {
+          return true;
+        }
+        if (object == null || getClass() != object.getClass()) {
+          return false;
+        }
+
+        Impl interledgerAddressImpl = (Impl) object;
+
+        return value.equals(interledgerAddressImpl.value);
+      }
+
+      @Override
+      public int hashCode() {
+        return value.hashCode();
+      }
+
+      @Override
+      public String toString() {
+        return "Impl{"
```

Comment:
i just realized that i think all of our concrete implementations .toString have "Impl"  as the prefix, which doesnt really help - we shouldnt we have the interface name as a prefix, e.g. "InterLedgerAddress.Impl"